docs/source/NRunner.rst

```diff
@@ -0,0 +1,185 @@
+architecture and implementation.  Some of them are related to
+limitations found in the current implementation, that were found
+to be too hard to remove without major breakage.  Other reasons
+are closely related to missing features that are deemed important.
+
+For instance, these are the current limitations of the Avocado test
+runner:
+
+* Test execution limited to the same machine, given that the
+  communication between runner and test is a Python queue (the remote
+  runner plugins actually execute an Avocado Job remotely, with all
+  the overhead and complications that it brings)
+* Test execution is limited to a single test at a time (non-parallel)
+* Test processes are not properly isolated and can affect the test
+  runner (including the "UI")
+
+And these are some features which it's believed to be more easily
+implemented under a different architecture and implementation:
+
+* Remote execution
+* Different test execution isolation models provided by the test runner
+  (process, container, virtual machine)
+* Distributed execution of tests across a pool of any combination of
+  processes, containers, virtual machines, etc.
+* Parallel execution of tests
+* Optimized runners for a given environment and or test type (for
+  instance, a runner written in RUST to run tests written in RUST
+  in an environment that already has RUST installed but not much
+  else)
+* Notification of execution results to many simultaneous "status
+  servers"
+* Disconnected test execution, so that results can be saved to a
+  device and collected by the runner
+* Simplified and automated deployment of the runner component into
+  execution environments such as containers and virtual machines
+
+Concepts:
+=========
+
+Runnable
+--------
+
+A runnable is a description of an entity that can be executed and
+produce some kind of result.  The description is abstract on purpose.
+A simple and obvious candidate for filling the description is a
+standalone executable, such as the ones available on your ``/bin``
+directory.
+
+A runnable must declare its kind.  Using the previous example of
+standalone executables, those may be given the unique kind identifier
+such as ``exec``.
+
+Each runnable kind may require a different amount of information to be
+provided so that it can be instantiated.  Using standalone executables
+as an example, the information required should be limtied to the
+location of the the standalone executable file.  The following
+pseudo-code may help to put these ideas together::
+
+  runnable_instance = create_runnable('exec', uri='/bin/true')
+
+Runner
+------
+
+A runner is an active entity that acts on the information of a
+runnable.  A runner will usually be tied to an specific kind of
+runnable, and will be able to able to act upon the specific
+information that runnable kind provides.
+
+The following pseudo-code may help to illustrate that::
+
+  runnable_instance = create_runnable('exec', path='/bin/sleep')
```

Comment:
let's use 'uri' as in test cases and other places in the doc.

Reply:
Yes, good catch.